[BUG, ENH, MRG] Tename clustering that uses F-tests to F_obs and refactor to allow time-frequency spatiotemporal clusters

alexrockhill opened the pull request:
Cleans up leftovers from https://github.com/mne-tools/mne-python/pull/10393.

So the first thing was that the default test was an F-test for the non-1-sample tests and so it should be called `F_obs` which gets output. That was actually done for the spatiotemporal but not the regular cluster test already so I made that consistent.

The next thing is that it really wasn't hard at all to add support for an arbitrary array shape of features (i.e. (times,) or (times, frequencies) or even (times, frequencies, whatever else)). Not that people should be using more than time-frequency dimensions, but its more general written this way. I do think spatiotemporal time-frequency clustering should be supported directly (rather than collapsing the time-frequency array before passing it as an argument) and this PR adds that.

alexrockhill commented:
So looks like the tests pass, and I think this is good but the one point I'm not sure on is whether the example should do something like this

```
# First we need to slightly modify the ANOVA function to be suitable for
# the clustering procedure. Also want to set some defaults.
# Let's first override effects to confine the analysis to the interaction
effects = 'A:B'

# %%
# A stat_fun must deal with a variable number of input arguments.
# Inside the clustering function each condition will be passed as flattened
# array, necessitated by the clustering procedure. The ANOVA however expects an
# input array of dimensions: subjects x conditions x observations (optional).
# The following function catches the list input and swaps the first and
# the second dimension and finally calls the ANOVA function.


def stat_fun(*args):
    return f_mway_rm(np.swapaxes(args, 1, 0), factor_levels=factor_levels,
                     effects=effects, return_pvals=False)[0]
```

for the `stat_fun` when spatiotemporal clustering over time frequencies. I guess it should be possible to leverage the repeated measures anova to account for the correlation in time, but I'm not really sure what's going on with `args` in this case. Maybe @dengemann  or @larsoner would have two minutes to weigh in on whether this is appropriate in this case. If not, I'll see if I can figure it out tomorrow.

agramfort commented on tutorials/stats-sensor-space/75_cluster_ftest_spatiotemporal.py:
```diff
@@ -179,6 +185,104 @@
+
+    # clean up viz
+    mne.viz.tight_layout(fig=fig)
+    fig.subplots_adjust(bottom=.05)
```
this block of code is neat but to me quite scary for new comers. We really need a high level API for stats and plotting stats results...

alexrockhill replied:
Those are mostly copy pastes from above in that file but I agree with your point

alexrockhill replied:
I might prefer to add a plotting function in another PR, it's a bit tricky because any plotting involves collapsing across either spatial or time-frequency dimensions and I think providing high-level code that does that may not be a great idea because then it's not as obvious what's being averaged/maxed.

I need to fix a DRY spot in cluster_level.py though before merging, I'll push that in minute.

dengemann replied:
we have had this code online like this now for > 6 years ... it should go in some plotting function at some point.

agramfort reviewed (commented):
@alexrockhill can you paste URLs to the rendered doc pages?

thx

alexrockhill commented:
https://42239-1301584-gh.circle-artifacts.com/0/dev/auto_tutorials/stats-sensor-space/75_cluster_ftest_spatiotemporal.html

alexrockhill commented:
I think the issue of the flattened time frequency dimensions in clustering can be solved with the `combine_adjacency` function which I just added. I think this PR works well and fixes inconsistencies in accepted inputs and documented outputs for the cluster functions.

Maybe someone has a nice explanation of the `stat_fun` swapaxes in the rmANOVA but I'm not changing it with this PR so I think that's a separate issue.

dengemann commented:
>Maybe @dengemann or @larsoner would have two minutes to weigh in on whether this is appropriate in this case. If not, I'll see if I can figure it out tomorrow.

I am not sure how you would go about this. You would e.g. create 5 time levels like 100ms bins as factors and then hope to test against the null that all time points are equal? You probably could do that but it's a different type of approach. The actual repeat measures we typically have in mind are multiple conditions etc.

alexrockhill commented:
> > Maybe @dengemann or @larsoner would have two minutes to weigh in on whether this is appropriate in this case. If not, I'll see if I can figure it out tomorrow.
> 
> I am not sure how you would go about this. You would e.g. create 5 time levels like 100ms bins as factors and then hope to test against the null that all time points are equal? You probably could do that but it's a different type of approach. The actual repeat measures we typically have in mind are multiple conditions etc.

Yeah that's what I figured, I don't think rmANOVA is appropriate in this case. I think it's just defining and appropriate `adjacency` parameter and then flattening all the dimensions that are not comparisons or conditions. That seems to work well in https://42264-1301584-gh.circle-artifacts.com/0/dev/auto_tutorials/stats-sensor-space/75_cluster_ftest_spatiotemporal.html?highlight=75_cluster (last section). Does that look reasonable to you?

dengemann commented:
looks plausible/reasonable!

alexrockhill commented:
Ok, sorry, the reshaping was already done for you and the code was in really good shape, it just took me a minute to get up to speed so I removed pretty much all changes to the code. There was just one small bug where, when excluding sensor positions, it was assumed that the shape was 3D and didn't support 4D in spatio_temporal_clustering so I fixed that.

This PR fixes that bug, makes it clear where an F statistics is returned as a default, fixes misleading images where F-statistics are plotted with a diverging colormap when they are strictly positive and standardizes documentation for the shape of data that can be passed into and is returned from the cluster test.

They are all minor changes but I think it's relatively important to get merged soon as I was mislead into thinking I had a T-statistic when I had an F-statistic and, because you lose the sign `(F=t**2)` for a comparison between two cases, that would be good to fix to avoid confusion.

alexrockhill commented:
I'm done making changes, this is good to go by me

larsoner reviewed (approved):
Okay with me, @agramfort feel free to merge if you're okay with the additional example complexity
